lowcar/devices/KoalaBear/KoalaBear.cpp

```diff
@@ -189,26 +224,37 @@ size_t KoalaBear::device_write(uint8_t param, uint8_t* data_buf) {
 
         // Params for Motor B
         case DUTY_CYCLE_B:
-            this->pid_enabled_b = FALSE;  //remove later for PID functionality
             this->target_speed_b = ((float*) data_buf)[0];
+			// limit to range [-1.0, 1.0]
+			this->target_speed_b = (this->target_speed_b > 1.0) ? 1.0 : this->target_speed_b;
+			this->target_speed_b = (this->target_speed_b < -1.0) ? -1.0 : this->target_speed_b;
             return sizeof(float);
         case DEADBAND_B:
             this->deadband_b = ((float*) data_buf)[0];
             return sizeof(float);
-        case CURRENT_B:
-            break;
-        case PID_ENABLED_B:
-            this->pid_enabled_b = data_buf[0];
+	    case INVERT_B:
+			this->invert_b = data_buf[0];
+			return sizeof(uint8_t);
+        case MANUAL_ENABLED_B:
+            this->manual_enabled_b = data_buf[0];
             return sizeof(uint8_t);
         case PID_KP_B:
-            this->pid_b->set_coefficients(((double*) data_buf)[0], this->pid_b->get_ki(), this->pid_b->get_kd());
+			// ignore executor setting all coefficients to 0 when going into idle
+			if (((double*) data_buf)[0] != 0.0) {
```

Comment:
This is the same as saying "ignore any 'zero' writes to the PID coefficients?"

Reply:
Ok so this was my (maybe) temporary solution to a problem I was struggling to solve. So Ashwin made it so that `executor` writes 0 (`int`s), 0.0 (`float`s), and False (`bool`s) to all writeable parameters when the run mode is switched to `IDLE` or when `executor` crashes. This is almost always the right thing to do, but it's definitely *not* the right thing to do for the PID coefficients (and one could argue it's not the right thing to do for the encoder values either). So without this here, currently what happens is that when you plug in the KoalaBear and the device enables, the PID coefficients are correct and set to the default values, and when you put the robot in auto or teleop, it works. But then when you switch it to `IDLE` all the motors stop working because the PID is by default enabled and all the coefficients were set to 0.

Putting these checks here solves the problem but it also makes it so that you can't write a 0 to, say KI--the device will literally just reject it. That's actually something a student might want to do (just have KP and no KI or KD), so it could be a problem. However, it's not something they're *likely* to do... if they're trying to set coefficients to zero you're supposed to just disable PID, but this solution is not the best.

Additionally, the reason why I had to change the parameter `pid_enabled_a` to `manual_enabled_a` is because of the very same issue. If I kept it to `pid_enabled_a`, `executor` would simply disable it immediately when going to `IDLE`, and from the KoalaBear's point of view, there's no way of telling whether that disable was from `executor` switching run modes or from the student actually disabling PID. I'd prefer to keep it as `pid_enabled_a` and not `manual_enabled_a` because `pid_enabled_a` sounds a bit more "correct"--like you're enabling a feature and then disabling it for more control.... does that make sense?

Anyways, I'm not sure what the best way to fix this problem is.... what do you think?

Reply:
If I'm getting this straight, executor will "zero" out write-able parameters on `IDLE` or when executor crashes. This is the right move for most cases, except for when we have non-zero default values? This is the first time we're encountering this case

Reply:
Something I'm thinking about is that in the lowcar code, we can define default values if necessary and executor will send a new `DEVICE_RESET` message (in addition to our other lowcar message types). When a lowcar device receives this message, it can write to parameters/pins on its own as appropriate. This would mean defining `Device::device_reset()`. Maybe this has to do with how we used to have a `Device::device_disable()`?? Unsure whether "disable" or "reset" is the more appropriate wording here.

Reply:
That's right, what you said in the first comment is exactly what's happening.
Do you think it's worth it to add a whole new lowcar message for this purpose? You're right, I think we can perhaps use this opportunity to reimagine the purpose of `device_disable()` in `lowcar`.... let's talk about this and I'll write up the results of the discussion in a later reply.

Reply:
Solution:
Change `executor` to only reset the params that it needs to, i.e. only the `velocity_a` and `velocity_b` params on the KoalaBear as of now. (Much simpler than expected! 👍 )